### PR TITLE
Architecture: Refactor Invocation

### PR DESCRIPTION
**Pull request by talekhinezh**
## Summary
Refactor Invocation Logic

## Details
* Simplify `call_method` by moving Root Global Reference logic into `CallFrame`. That is, rather than recomputing the global node of a new actor, we just keep track of the global node of every node reference.
* Remove `Visibility::Actor`
* Generify Actor

### CODE REVIEW

File: radix-engine/src/kernel/actor.rs
Comment (iamyulong) [inline radix-engine/src/kernel/actor.rs:18]: Oh yeah?

File: radix-engine/src/kernel/call_frame.rs
Comment (iamyulong) [inline radix-engine/src/kernel/call_frame.rs:26]: What's the relationship between `copy_references`, `copy_transient_references` and `copy_direct_references`?

File: radix-engine/src/kernel/call_frame.rs
Comment (iamyulong) [inline radix-engine/src/kernel/call_frame.rs:70]: This is probably a naming thing, but `RootNodeType` abstraction doesn't look right, and I think we're mixing reference type with node location.

File: radix-engine/src/kernel/actor.rs
Comment (iamyulong) [inline radix-engine/src/kernel/actor.rs:106]: We should use `InternalAddress` for transient and direct access references, as we're using `GlobalAddress` for global references?

**Review by iamyulong — COMMENTED** [review]
Nice work to address many tech debt.

Although, I'd suggest we remove premature `RootNodeType` abstraction.

File: radix-engine/src/kernel/actor.rs
Comment (talekhinezh) [reply radix-engine/src/kernel/actor.rs:73]: As discussed, "actor" is the wrong abstraction as the CallFrame has no notion of "actor".

File: radix-engine/src/kernel/call_frame.rs
Comment (iamyulong) [inline radix-engine/src/kernel/call_frame.rs:60]: Should split `value.references()` in case a function returns a direct access references.

File: radix-engine/src/kernel/call_frame.rs
Comment (iamyulong) [inline radix-engine/src/kernel/call_frame.rs:67]: Can we rename to `add_copy_global_reference`

File: radix-engine/src/kernel/call_frame.rs
Comment (iamyulong) [inline radix-engine/src/kernel/call_frame.rs:169]: Can we add a TODO here that we'd like to change the return to `Vec<ReferenceOrigin>`?

This "global supersedes direct access reference" is very arbitrary and unsound.

File: radix-engine/src/kernel/call_frame.rs
Comment (iamyulong) [inline radix-engine/src/kernel/call_frame.rs:233]: Nice!!!!

File: radix-engine/src/kernel/call_frame.rs
Comment (iamyulong) [inline radix-engine/src/kernel/call_frame.rs:417]: Panic? Can it be triggered by clients?

File: radix-engine/src/kernel/call_frame.rs
Comment (iamyulong) [inline radix-engine/src/kernel/call_frame.rs:420]: Name: `reference_origin`

File: radix-engine/src/kernel/call_frame.rs
Comment (iamyulong) [inline radix-engine/src/kernel/call_frame.rs:432]: Ditto

File: radix-engine/src/kernel/call_frame.rs
Comment (iamyulong) [inline radix-engine/src/kernel/call_frame.rs:464]: Ditto

File: radix-engine/src/kernel/kernel.rs
Comment (iamyulong) [inline radix-engine/src/kernel/kernel.rs:185]: `CallFrameData` does sound better!

**Review by iamyulong — APPROVED** [review]
Some minor clean-ups are still required.

Approving this PR, assuming you will fix them before merge.

File: radix-engine/src/kernel/call_frame.rs
Comment (talekhinezh) [reply radix-engine/src/kernel/call_frame.rs:60]: good catch

File: radix-engine/src/kernel/call_frame.rs
Comment (talekhinezh) [reply radix-engine/src/kernel/call_frame.rs:417]: Nope, can never be triggered by clients as this is controlled by the system layer. You can see that `copy_to_stable_transient_references` is empty in `CallFrameMessage::from_output()`

**Pull request merged**
